src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php

```diff
@@ -110,6 +111,11 @@ public function validate($value, $constraints = null, $groups = null)
         $previousMetadata = $this->context->getMetadata();
         $previousPath = $this->context->getPropertyPath();
         $previousGroup = $this->context->getGroup();
+        $previousConstraint = null;
+
+        if ($this->context instanceof ExecutionContext || method_exists($this->context, 'getConstraint')) {
```

Comment:
I did not test it, but the `instanceof` check should be faster and should be true most of the time.